### PR TITLE
Customer Support Sample

### PR DESCRIPTION
**Pull request by luisquintanilla**

### CODE REVIEW

**Comment by luisquintanilla** [comment]
@jmatthiesen this is ready for review when you get a chance. Thanks.

File: .devcontainer.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+                "ms-dotnettools.dotnet-interactive-vscode"
+            ]
+        }
+    },
+    "runArgs": [
+        "--gpus", "all"
+    ],
+    "postCreateCommand": "curl -fsSL https://ollama.com/install.sh | sh",
+    "postStartCommand": "ollama serve",
+    "containerEnv": {
```
Comment (luisquintanilla) [inline .devcontainer.json:28]: Remove these.

File: src/samples/CustomerSupport/README.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
+# Customer Support 
+
```
Comment (luisquintanilla) [inline src/samples/CustomerSupport/README.md:2]:
```suggestion
Add project description.
```

**Comment by jmatthiesen** [comment]
This is looking great, thanks for sharing, Luis!

Per a conversation we just had IRL, here are some suggestion
* Update the Readme to provide an overview of the scenario being demonstrated with this sample.
* We want to establish a pattern for folder structure to demonstrate different use cases in this repo. For a folder name, what do you think of `src/data-ingestion/` or `src/data-ingestion-localdotnet`? Other suggestions welcome

File: .devcontainer.json (hunk 1 above)
Comment (luisquintanilla) [reply .devcontainer.json]: Done

File: src/samples/CustomerSupport/README.md (hunk 2 above)
Comment (luisquintanilla) [reply src/samples/CustomerSupport/README.md:2]: Done

**Comment by luisquintanilla** [comment]
I think since this one is more generally concerned with chat, we can use chat for this one.

So the path can be `src/chat`. We can play around with it after the fact.

File: .devcontainer.json [hunk 3]
```diff
@@ -0,0 +1,28 @@
+{
```
[the remark this answers is not in the document]
Comment (jmatthiesen) [reply .devcontainer.json:1]: I'm also getting this error when using Codespaces:

`docker: Error response from daemon: could not select device driver "" with capabilities: [[gpu]].
2024-08-17 22:51:40.087Z: Stop: Run: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,src=/var/lib/docker/codespacemount/workspace,dst=/workspaces --mount type=volume,src=dind-var-lib-docker-0hcbhh2c7vldoj773drm1bjldnb89u0rt96sl9nju22d9ou8d14n,dst=/var/lib/docker --mount source=/root/.codespaces/shared,target=/workspaces/.codespaces/shared,type=bind --mount source=/var/lib/docker/codespacemount/.persistedshare,target=/workspaces/.codespaces/.persistedshare,type=bind --mount source=/.codespaces/agent/mount,target=/.codespaces/bin,type=bind --mount source=/mnt/containerTmp,target=/tmp,type=bind --mount type=bind,src=/.codespaces/agent/mount/cache,dst=/vscode -l Type=codespaces -e CODESPACES=******** -e ContainerVersion=13 -e RepositoryName=ai-samples --gpus all --label ContainerVersion=13 --hostname codespaces-82c536 --add-host codespaces-82c536:127.0.0.1 --cap-add sys_nice --network host --privileged --entrypoint /bin/sh vsc-ai-samples-ac7d8fa2041ac73b151cb3eae118357442b9086c698dd3a2853cd6402e799649-features -c echo Container started
2024-08-17 22:51:40.119Z: Error: Command failed: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,src=/var/lib/docker/codespacemount/workspace,dst=/workspaces --mount type=volume,src=dind-var-lib-docker-0hcbhh2c7vldoj773drm1bjldnb89u0rt96sl9nju22d9ou8d14n,dst=/var/lib/docker --mount source=/root/.codespaces/shared,target=/workspaces/.codespaces/shared,type=bind --mount source=/var/lib/docker/codespacemount/.persistedshare,target=/workspaces/.codespaces/.persistedshare,type=bind --mount source=/.codespaces/agent/mount,target=/.codespaces/bin,type=bind --mount source=/mnt/containerTmp,target=/tmp,type=bind --mount type=bind,src=/.codespaces/agent/mount/cache,dst=/vscode -l Type=codespaces -e CODESPACES=true -e ContainerVersion=13 -e RepositoryName=ai-samples --gpus all --label ContainerVersion=13 --hostname codespaces-82c536 --add-host codespaces-82c536:127.0.0.1 --cap-add sys_nice --network host --privileged --entrypoint /bin/sh vsc-ai-samples-ac7d8fa2041ac73b151cb3eae118357442b9086c698dd3a2853cd6402e799649-features -c echo Container started
2024-08-17 22:51:40.120Z: trap "exit 0" 15
2024-08-17 22:51:40.120Z: {"outcome":"error","message":"Command failed: docker run --sig-proxy=false -a STDOUT -a STDERR --mount type=bind,src=/var/lib/docker/codespacemount/workspace,dst=/workspaces --mount type=volume,src=dind-var-lib-docker-0hcbhh2c7vldoj773drm1bjldnb89u0rt96sl9nju22d9ou8d14n,dst=/var/lib/docker --mount source=/root/.codespaces/shared,target=/workspaces/.codespaces/shared,type=bind --mount source=/var/lib/docker/codespacemount/.persistedshare,target=/workspaces/.codespaces/.persistedshare,type=bind --mount source=/.codespaces/agent/mount,target=/.codespaces/bin,type=bind --mount source=/mnt/containerTmp,target=/tmp,type=bind --mount type=bind,src=/.codespaces/agent/mount/cache,dst=/vscode -l Type=codespaces -e CODESPACES=true -e ContainerVersion=13 -e RepositoryName=ai-samples --gpus all --label ContainerVersion=13 --hostname codespaces-82c536 --add-host codespaces-82c536:127.0.0.1 --cap-add sys_nice --network host --privileged --entrypoint /bin/sh vsc-ai-samples-ac7d8fa2041ac73b151cb3eae118357442b9086c698dd3a2853cd6402e799649-features -c echo Container started\ntrap \"exit 0\" 15\n/usr/local/share/docker-init.sh\nexec \"$@\"\nwhile sleep 1 & wait $!; do :; done -","description":"An error occurred setting up the container.","containerId":"467320ac6b3e85493a22fbd26e043012f0dca8e6b1217c64f693a0524e817e6f"}
`

**Review by jmatthiesen — CHANGES_REQUESTED** [review]

File: src/chat/CustomerSupport/OllamaChatCompletionService.cs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+using System.Text.Json;
+using Azure.AI.OpenAI;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using OllamaSharp;
+using OllamaSharp.Models.Chat;
+using Spectre.Console;
+
+public class OllamaChatCompletionService : IChatCompletionService
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/chat/CustomerSupport/OllamaChatCompletionService.cs:11]: The intent of this is mainly aesthetic for now. It's a simplified implementation of the Ollama feature branch in SK. 

In the future, we'd replace this with the SK implementation.

File: src/chat/CustomerSupport/Program.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+using System.Reflection.Metadata.Ecma335;
+using static Utils;
+using Microsoft.SemanticKernel.Embeddings;
+// using Microsoft.SemanticKernel.Embeddings;
+
+// Configure AI
+var apiKey = "";
+var endpoint = "http://localhost:11434/";
+
+var openAIKey = Environment.GetEnvironmentVariable("AZURE_AI_KEY");
```
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/chat/CustomerSupport/Program.cs:23]: Do you have a sample that shows how to do that?

File: .devcontainer.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply .devcontainer.json:1]: I'll create separate versions (CPU and GPU)

[the remark this answers is not in the document]
Comment (luisquintanilla) [reply .devcontainer.json]: Done

File: src/chat/CustomerSupport/Program.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (luisquintanilla) [reply src/chat/CustomerSupport/Program.cs]: Pushed managed identity in latest update.

**Comment by luisquintanilla** [comment]
@jmatthiesen this is ready for review / merge with the changes you suggested.

**Review by jmatthiesen — APPROVED** [review]

**Comment by luisquintanilla** [comment]
Thanks for the review @jmatthiesen.

Can you please merge when you get a chance. I don't have write access. Thanks!

**Pull request merged**